novoda-blog-diversity/lib/data-source.js

```diff
@@ -19,97 +19,49 @@ const parseRss = (url) => {
 }
 
 const toViewState = (items) => {
-    var mentions = new Map([["android", 0], ["ios", 0], ["cross-platform", 0]])
-    mentions.set("android", counUniqueTagsIn(items, androidTags))
-    mentions.set("ios", counUniqueTagsIn(items, iosTags))
-    mentions.set("cross-platform", counUniqueTagsIn(items, crossPlatformTags))
-        
+    var map = tagsWithUsagesFrom(items)
     return {
-        mostUsedTags: mostUsedTagsFrom(items),
-        crossPlatformMentions: mentions.get("cross-platform"),
-        androidMentions: mentions.get("android"),
-        iosMentions: mentions.get("ios")
+        mostUsedTags: mostUsedTagsFrom(map),
+        crossPlatformMentions: counUniqueTagsIn(map, crossPlatformTags),
+        androidMentions: counUniqueTagsIn(map, androidTags),
+        iosMentions: counUniqueTagsIn(map, iosTags)
     }
 }
 
-function mostUsedTagsFrom(items){
-    return sortMapDescendingByValue(tagsWithUsagesFrom(items))
-        .splice(0,3)
-        .map(tagUsage => tagUsage[0])
+function mostUsedTagsFrom(map) {
+    return Object.keys(map)
+        .map(key => [key, map[key]]) // transforms map into array
+        .sort((firstTag, secondTag) => secondTag[1] - firstTag[1]) // sort array by value
+        .splice(0, 3) // take the first 3 elements
+        .map(tagUsage => tagUsage[0]) // take only the names
         .join(", ")
 }
 
-function tagsWithUsagesFrom(items){
+function tagsWithUsagesFrom(items) {
     var tagsWithUsages = new Map()
 
     items
         .filter(item => item.categories != null)
-        .map(function(item){ return item.categories})
-        .map(function(categories){
-            categories.forEach(category => {
-                if (tagsWithUsages[category] == undefined){
-                    tagsWithUsages[category] = 0
-                } 
-                
-                tagsWithUsages[category] = tagsWithUsages[category]+1                 
-            });
-        })
-
-    return tagsWithUsages
-}
-
-function sortMapDescendingByValue(obj)
-{
-	var sortable=[];
-	for(var key in obj)
-		if(obj.hasOwnProperty(key))
-			sortable.push([key, obj[key]]);
-	
-	sortable.sort(function(a, b)
-	{
-		var x=a[1],
-			y=b[1];
-		return y<x ? -1 : y>x ? 1 : 0;
-    });
-    
-    return sortable;
-}
-
-function counUniqueTagsIn(items, tags){
-    var mentionCount = 0
+        .map(item => item.categories)
+        .reduce((prev, curr) => prev.concat(curr))
```

Comment:
does this work? I would expect the reduce to need a initial value of `[]`

Reply:
ahhh it's an array of arrays, this means you're adding the first element twice.

```
Value to use as the first argument to the first call of the callback. If no initial value is supplied, the first element in the array will be used. Calling reduce() on an empty array without an initial value is an error.
```

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce